docs/package_templates/autotools_package/all/conanfile.py

```diff
@@ -75,64 +85,99 @@ def requirements(self):
 
     def validate(self):
         # validate the minimum cpp standard supported. Only for C++ projects
-        if self.info.settings.compiler.cppstd:
+        if self.info.settings.compiler.get_safe("cppstd"):
             check_min_cppstd(self, 11)
         if self.info.settings.os not in ["Linux", "FreeBSD", "MacOS"]:
             raise ConanInvalidConfiguration(f"{self.ref} is not supported on {self.info.settings.os}.")
 
     # if another tool than the compiler or autotools is required to build the project (pkgconf, bison, flex etc)
     def build_requirements(self):
+        # only if we have to call autoreconf
         self.tool_requires("libtool/x.y.z")
-        self.tool_requires("pkgconf/x.y.z")
+        # only if upstream configure.ac relies on PKG_CHECK_MODULES macro
+        if not self.conf.get("tools.gnu:pkg_config", default=False, check_type=str):
+            self.tool_requires("pkgconf/x.y.z")
+        # required to suppport windows as a build machine
+        if self._settings_build.os == "Windows":
+            self.win_bash = True
+            if not self.conf.get("tools.microsoft.bash:path", default=False, check_type=str):
+                self.tool_requires("msys2/cci.latest")
+        # for msvc support to get compile & ar-lib scripts (may be avoided if shipped in source code of the library)
+        if is_msvc(self):
+            self.tool_requires("automake/x.y.z")
 
     def source(self):
         get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
 
     def generate(self):
-        # autotools usually uses 'yes' and 'no' to enable/disable options
-        yes_no = lambda v: "yes" if v else "no"
+        # inject tools_requires env vars in build scope (not needed if there is no tool_requires)
+        env = VirtualBuildEnv(self)
+        env.generate()
+        # inject requires env vars in build scope
+        # it's required in case of native build when there is AutotoolsDeps & at least one dependency which might be shared, because configure tries to run a test executable
+        if not cross_building(self):
+            env = VirtualRunEnv(self)
+            env.generate(scope="build")
         # --fpic is automatically managed when 'fPIC'option is declared
         # --enable/disable-shared is automatically managed when 'shared' option is declared
         tc = AutotoolsToolchain(self)
-        tc.configure_args.append("--with-foobar=%s" % yes_no(self.options.with_foobar))
-        tc.configure_args.append("--enable-tools=no")
-        tc.configure_args.append("--enable-manpages=no")
+        # autotools usually uses 'yes' and 'no' to enable/disable options
+        yes_no = lambda v: "yes" if v else "no"
+        tc.configure_args.extend([
+            f"--with-foobar={yes_no(self.options.with_foobar)}",
+            "--enable-tools=no",
+            "--enable-manpages=no",
+        ])
         tc.generate()
```

Comment:
In #13355, I've just figured out that to get the compiler wrapper script to work, I had to pass in the Environment as an argument to the AutotoolsToolchain `generate()` function. Perhaps generating the toolchain after the Environment for the compiler wrapper has been set destroys the compiler wrapper variables. I think that's worth mentioning.

Reply:
Yes, the order in which generators are called my be important. Basically you want to call `Virtual[Build/Run]Env` first, then other generators, then specific `Environment`.

I've proposed https://github.com/conan-io/conan/pull/12194 for robust handling of Visual Studio in AutotoolToolchain, and honor in a robust way injection of CC, CXX etc from profile, but it deserves additional testing in conan client, and maybe a slightly different implementation (maybe a separate generator). It also suggest to add specific `conf` parameters to provide a canonical conan way to inject compiler path for users.